Monitoring setup for the LSST messaging stack at RAL

Neonbluestoplight opened the pull request:
A full description of the monitoring stack along with the setup of its components can be obtained from [this confluence page](https://stfc.atlassian.net/wiki/spaces/RO/pages/961151089/Prometheus+JMX+Exporter+for+Kafka+monitoring). The summary of the stack is given below:

1. Prometheus JMX exporter attaches to Kafka, MM2 and Zookeeper and sends configured resource level metrics to Prometheus.
2. The [danielqsj/kafka_exporter](https://github.com/danielqsj/kafka_exporter) sends high level metrics like consumer lag to Prometheus as well
3. Grafana uses Prometheus as a data source to visualise Kafka metrics

<img width="2640" height="1253" alt="image" src="https://github.com/user-attachments/assets/fc52d88b-3d87-4823-8075-9aaad572cbc2" />

A working example of the dashboard itself can be found in http://172.16.102.73:3000/ (accessed via the VPN). However, this working example has IngestD disabled to avoid possible redundant behaviour so consumer lag metrics will not be reported until the merge is complete

Neonbluestoplight committed:
Basic monitoring implementation

Added services like grafana, prometheus and alertmanager for developing a monitoring workflow in a directory titled "monitoring". This directory has its own docker-compose.yaml which is linked to the main one so that they can all run together.

Neonbluestoplight committed:
Adding jmx prometheus exporter

The kafka-exporter only provides high level metrics (e.g. consumer groups, partition info etc.) but does not provide info like messages per minute. So we are using the prometheus jmx exporter plugin for this.

Neonbluestoplight committed:
Add JMX exporter jar file

Using the jar file since it is easier to work with and cleaner.

Neonbluestoplight committed:
Adding memory caps to the JMX config

Neonbluestoplight committed:
Changes to ensure JMX exporter scrapes metrics

- Changed the name of the jmx exporter directory to be compatible with the javaagent path
- Changed port mapping in some of the javaagent paths to ensure consistency
- Disabled kafka broker healthcheck because it was not running and causing mirrormaker to fail
- Added more information in JMX config files based on https://github.com/prometheus/jmx_exporter/blob/main/examples/kafka-2_0_0.yml
- Updated grafana dashboard to scrape hardware related metrics

Neonbluestoplight committed:
Moved kafka-exporter to monitoring compose file

Moved the kafka-exporter service to the monitoring docker compose to keep all of the monitoring components in the same file.

Neonbluestoplight committed:
Dashboard updates

Updates to the tile sizing of the panels and the data extracted

Neonbluestoplight committed:
Added comments on the monitoring config files

For more context, added comments to the config files. For the monitoring docker compose, I followed the same format as the main docker compose for consistency

Neonbluestoplight committed:
Dashboard updates

- Added some descriptions for the panels.
- Ensured the legends are readable by only displaying topic name
- Included filters to remove unnecessary topics (like kafka internal topics)

Neonbluestoplight committed:
Dashboard updates - Filters

Added filters per topic for panels like Messages in per topic

Neonbluestoplight committed:
Update dashboard and change Kafka version

The dashboard was updaed to include more parameters such as disk speeds and change existing panels to be more in line with SLAC strimzi configuration ones.

Also changed the Kafka version to 7.9.0 to not intefere with KRaft changes.

Neonbluestoplight committed:
Dashboard changes

Added a few metrics that might help detect blips since there is currently a difficulty detecting any sort of blips in messages received. Also added some consumer group metrics to prepare for the eventual merge with the official monitoring stack.

Neonbluestoplight committed:
Moved the Grafana panels around a bit

Neonbluestoplight committed:
Commenting out the alertmanager setup.

alertmanager is most likely not a necessary sevice to have but it was also somewhat setup so I have commented out any refernces so it doesn't interrupt the running of containers but can be setup relatively quickly when needed.

Neonbluestoplight committed:
Commented out ingestd for the time-being

To allow for final tests before merge into main

Neonbluestoplight committed:
Changed heap memory for Kafka, MM2, Zookeeper

To work well with moderate to high throughputs, changed the heap memory to the values below:
Kafka     : initial - 3G, max - 3G
MM2       : initial - 1G, max - 1G
Zookeeper : initial - 1G, max - 1G

Neonbluestoplight committed:
Updated version of Prometheus JMX Exporter

- Updated the Prometheus JMX Exporter version from 1.2.0 to 1.3.0
- Also commented IngestD back in for the merge request

thysk commented on docker-compose.yaml:
```diff
@@ -84,8 +96,8 @@ services:
         condition: service_healthy
-      broker1:
-        condition: service_healthy
+      # broker1:
```
Is there a reason you disabled this?

Neonbluestoplight replied:
At the time, the healthchecks were failing due to an unkown reason so I had disabled them since they did not affect the overall functioning of the system. The cause of the failiures was determined to be the linking of the Prometheus JMX exporter using `KAFKA_OPTS` which was preventing any of the kafka related commands from running within the container (which was required for the healthcheck).

This has been changed to using `EXTRA_ARGS` now for consistency with the Zookeeper and MM2 containers as well.

thysk commented on docker-compose.yaml:
```diff
@@ -28,15 +38,12 @@ services:
-      timeout: 10s
-      retries: 3
+      # No healthcheck for MirrorMaker 2
+      test: ["NONE"]
```
is there a reason for stopping these tests?

Neonbluestoplight replied:
It was decided that tests on MM2 were not strictly necessary as the only role of MM2 would be to redirect USDF messages to the kafka cluster, which would be recorded by monitoring anyways.

thysk commented on monitoring/alertmanager/alertmanager.yml:
Is there a point to this file if it is all commented out?

Neonbluestoplight replied:
I have removed the file and any references to alertmanager now

thysk commented on monitoring/docker-compose.yaml:
```diff
@@ -0,0 +1,104 @@
+
+services:
+  # Grafana dashboard for monitoring Kafka data via Prometheus
+  grafana:
```
As we are going to register the metrics in Vande we no longer need to deploy our own grafana

thysk commented on monitoring/grafana/dashboards/main.json:
Good to keep a record of this (and as up to date as you can going forwards), but this will not be needed by anything in the stack, but a good fallback should Vande forget the dashboard or another person wants it for another use

Neonbluestoplight replied:
I have kept an updated copy of this for use of Vande and removed any references to Grafana in the stack

thysk commented on monitoring/grafana/provisioning/dashboards/dashboard.yaml:
Good to keep a record of this (and as up to date as you can going forwards), but this will not be needed by anything in the stack, but a good fallback should Vande forget the dashboard or another person wants it for another use

thysk commented on monitoring/grafana/provisioning/datasources/datasources.yml:
Good to keep a record of this (and as up to date as you can going forwards), but this will not be needed by anything in the stack, but a good fallback should Vande forget the dashboard or another person wants it for another use

thysk commented on monitoring/jmx_exporter/mirrormaker_jmx_config.yml:
is this needed? as it is currently commented out, and says it will eventually be used for health checks

Neonbluestoplight replied:
I have added mirrormaker related monitoring from [Prometheus JMX exporter](https://github.com/prometheus/jmx_exporter/blob/main/examples/kafka-connect.yml) in case we want relevant metrics for monitoring now.

thysk commented on monitoring/prometheus/prometheus.yml:
```diff
@@ -0,0 +1,23 @@
+scrape_configs:
+  - job_name: 'prometheus'
+    static_configs:
+      - targets: ['localhost:9090']
```
for configuration consistency I believe this should be: - targets: ['prometheus:9090']
But please do test before accepting this suggestion

Neonbluestoplight replied:
Have changed and ensured it functions

thysk commented on docker-compose.yaml:
```diff
@@ -1,25 +1,35 @@
 ## Docker Compose File for Rubin-LSST DRP Messaging Stack
 
+# Including the docker compose from monitoring to run monitoring services
+include:
```
I like this as a way to separate out various aspects into different compose files but have a master one to deploy from

thysk reviewed (commented):
I have left some comments, please do respond to them, and make any changes to the PR you feel are needed.
As I mention in one of the comments, we should no longer need to deploy grafana as you have done the hard work with the dev instance and can move to using Vande as the visulaisation tool

thysk reviewed (commented):
I have left some comments, please do respond to them, and make any changes to the PR you feel are needed.
As I mention in one of the comments, we should no longer need to deploy grafana as you have done the hard work with the dev instance and can move to using Vande as the visulaisation tool

thysk reviewed (commented):
I have left some comments, please do respond to them, and make any changes to the PR you feel are needed.
As I mention in one of the comments, we should no longer need to deploy grafana as you have done the hard work with the dev instance and can move to using Vande as the visulaisation tool

thysk reviewed (changes requested):
I have left some comments, please do respond to them, and make any changes to the PR you feel are needed.
As I mention in one of the comments, we should no longer need to deploy grafana as you have done the hard work with the dev instance and can move to using Vande as the visulaisation tool

Neonbluestoplight committed:
Removed Alertmanager setup

Based on suggestions in the PR, Any references to alertmanager were removed since they were unnecessary.

Neonbluestoplight committed:
Healthchecks, MM2 JMX, datasource name changes

Added Healthcheck for Kafka which involved changing environment variables around a bit. Also added kafka-connect metrics in JMX to be thorough and changed the name of all the prometheus endpoints to work with Vande eventually.

Neonbluestoplight committed:
Removed Grafana & Docker compose cleanup

Grafana docker image and provisioning settings were removed since Vande Grafana is going to be the dashboard source now.

Also cleaned up the Docker compose file:
- Ensured healthchecks worked for Kafka and Zookeeper.
- Added "depends_on" tags to the rest of the settings to ensure a rigid startup order.
- Rearranged Docker compose variables to maintain consistency throughout the file.]

mathewcsims commented:
Hey @Neonbluestoplight — I've no feedback on this beyond what Tim's already given. Except to say thank you for picking this up and doing such a great job with it. Assuming @Thysk is happy, go ahead and merge it and test it for real :)

Also, if you want to add a slide about it to the GridPP presentation (and tell me what to say!) go ahead; you should get credit for the work you've done.

Neonbluestoplight commented:
> Hey @Neonbluestoplight — I've no feedback on this beyond what Tim's already given. Except to say thank you for picking this up and doing such a great job with it. Assuming @Thysk is happy, go ahead and merge it and test it for real :)
> 
> Also, if you want to add a slide about it to the GridPP presentation (and tell me what to say!) go ahead; you should get credit for the work you've done.

Thank you so much! I will merge it after a final confirmation from @Thysk and as for the slide, I will run it by you when I can, thank you for the opportunity!

thysk commented:
Looks good to me, was awaiting @mathewcsims approval and merge.
But as I see there is approval I will now merge
